### PR TITLE
Retry unification deprecations removal

### PR DESCRIPTION
**Pull request by tkountis**

### CODE REVIEW

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:126]: Consider using a private static constant for this default

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ClientClosureRaceTest.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ClientClosureRaceTest.java:185]: The snapshot build is currently failing with: https://github.com/apple/servicetalk/runs/4476095925
```
Check failure on line 184 in servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ClientClosureRaceTest.java 
Comment has incorrect indentation level 28, expected is 27, indentation should be the same level as line 185.
```

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java:86]: 1. This public constant misses javadoc.
2. Did you consider a public static method instead of a constant?

File: servicetalk-client-api/src/test/java/io/servicetalk/client/api/DefaultAutoRetryStrategyProviderTest.java
Comment (idelpivnitskiy) [inline servicetalk-client-api/src/test/java/io/servicetalk/client/api/DefaultAutoRetryStrategyProviderTest.java:59]: We shouldn't lose these test scenarios when we migrate users to the new filter

**Review by idelpivnitskiy — CHANGES_REQUESTED** [review]

File: servicetalk-client-api/src/test/java/io/servicetalk/client/api/DefaultAutoRetryStrategyProviderTest.java
Comment (tkountis) [reply servicetalk-client-api/src/test/java/io/servicetalk/client/api/DefaultAutoRetryStrategyProviderTest.java:59]: Good catch, re-added.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java:89]: With the constant, you have less flexibility to adjust it. With a static method it's easier to change what it returns.

**Review by idelpivnitskiy — APPROVED** [review]
LGTM after build passes

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java:115]: Is this even possible? a `create` method is invoked by the client builder, meaning nobody else has a reference to the filter to inject something

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java:241]: IIUC, you plan to cherry-pick the new approach into 0.41. Make sure that `AutoRetryStrategyProvider#DISABLE_AUTO_RETRIES` references this new method as a replacement

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java
Comment (tkountis) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java:115]: I will remove these, it was only for debugging. Regardless, the problem is that the shared DEFAULT is re-injected state in test environment and concurrent testing and this is causing the problem. The factory is holding state. This shouldn't be a problem for real use-cases but nevertheless let's discuss about it before we merge changes.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java:392]: With this approach I'm a bit concerned about number of wrapping layers created for each client. Because it will always be on request path, there is a risk it may affect JIT or some other JVM optimizations.

What I had in mind was to extend `StreamingHttpClientFilterFactory` API only for internal purposes:
```java
    @FunctionalInterface
    interface ExtendedStreamingHttpClientFilterFactory extends StreamingHttpClientFilterFactory {

        StreamingHttpClientFilter create(FilterableStreamingHttpClient client, Publisher<Object> lbEventStream, SdStatusCompletable sdStatus);
    }
```
Then we can use it to pass 2 extra arguments:
```java
    private static ExtendedStreamingHttpClientFilterFactory appendExtendedFilter(
            @Nullable final ExtendedStreamingHttpClientFilterFactory currClientFilterFactory,
            final StreamingHttpClientFilterFactory appendClientFilterFactory) {
        if (appendClientFilterFactory instanceof RetryingHttpRequesterFilter) {
            if (currClientFilterFactory == null) {
                return (client, lbEventStream, sdError) -> {
                    ContextAwareRetryingHttpRequesterFilter filter = (ContextAwareRetryingHttpRequesterFilter) ((RetryingHttpRequesterFilter) appendClientFilterFactory).create(client);
                    filter.inject(lbEventStream, sdError);
                    return filter;
                };
            } else {
                return (client, lbEventStream, sdError) -> {
                    ContextAwareRetryingHttpRequesterFilter filter = (ContextAwareRetryingHttpRequesterFilter) ((RetryingHttpRequesterFilter) appendClientFilterFactory).create(client);
                    filter.inject(lbEventStream, sdError);
                    return currClientFilterFactory.create(filter, lbEventStream, sdError);
                };
            }

        } else {
            if (currClientFilterFactory == null) {
                return (client, lbEventStream, sdError) -> appendClientFilterFactory.create(client);
            } else {
                return (client, lbEventStream, sdError) -> currClientFilterFactory.create(appendClientFilterFactory.create(client), lbEventStream, sdError);
            }
        }
    }
```
^ it also creates extra wrappers for `StreamingHttpClientFilterFactory`, but those will be only at the build time.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ContextAwareStreamingHttpClientFilterFactory.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ContextAwareStreamingHttpClientFilterFactory.java:27]: If this interface doesn't leak outside `DefaultSingleAddressHttpClientBuilder`, consider defining it there instead of the top level interface

**Review by idelpivnitskiy — APPROVED** [review]
Cherry-pick will be tricky :)

**Comment by tkountis** [comment]
Closed in favor of https://github.com/apple/servicetalk/pull/2019

**Pull request closed without merging**
